### PR TITLE
[AIRFLOW-4306] Global operator extra links

### PR DESCRIPTION
**Pull request by milton0825**
Make sure you have checked _all_ steps below.

### Jira

- [X] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-4306

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

A way to register global operator extra link that is available for all the operators through airflow plugin.

Screen Shot 2019-04-11 at 11 45 54 PM

### Tests

- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [X] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [X] Passes `flake8`

### CODE REVIEW

**Comment by Fokko** [comment]
Why would we need this? Can you give some examples or actual useful links that you would register here? I feel like that there are already too many buttons in the dialog.

**Comment by milton0825** [comment]
@Fokko I think it is up to the users to decide whether they need to add these extra links. One use case at Lyft is to add extra links to task logs in Kibana/Elastic search and we probably will add links to task logs that are archived in S3. Therefore we need a way other than https://github.com/apache/airflow/pull/5059 to register operator extra links at global level.

**Comment by milton0825** [comment]
PTAL @feng-tao @ashb

**Comment by milton0825** [comment]
PTAL @XD-DENG @Fokko

**Comment by feluelle** [comment]
I think it's a nice idea and can be kind of useful in some cases but I think I agree more with @Fokko.

**Comment by milton0825** [comment]
@Fokko @feluelle 
I am not too concerned that there will be too many buttons in the dialog UI. It is up to the user to decide whether they want to add more buttons. If the user thinks that they don't want those buttons, they simply don't register any. I would rather provide the more flexibility to the users than limit it.

Would like to hear more about the reason behind that against such flexibility.

**Comment by feng-tao** [comment]
@Fokko @feluelle , I agree with @milton0825 that it would be good to provide the flexibility instead limiting that. In fact we have been running this feature at Lyft in production for long time(https://eng.lyft.com/running-apache-airflow-at-lyft-6e53bb8fccff). E.g one use case could be make the kibana log url for all the operators as kibana log searching is much easier to look at the Airflow log.

**Comment by milton0825** [comment]
Tests passed in: https://travis-ci.org/milton0825/airflow/builds/522044293

**Comment by milton0825** [comment]
tests passed on: https://travis-ci.org/milton0825/airflow/builds/522373084

**Comment by feng-tao** [comment]
@Fokko , @feluelle happy to address any other concerns you guys have. Let us know.

**Comment by milton0825** [comment]
@Fokko @feluelle let us know if you have any concerns~

**Comment by feluelle** [comment]
@milton0825 @feng-tao actually I didn't know that we already had #5059 in. 

> I am not too concerned that there will be too many buttons in the dialog UI. It is up to the user to decide whether they want to add more buttons. If the user thinks that they don't want those buttons, they simply don't register any. I would rather provide the more flexibility to the users than limit it.

Thanks Chao-Han. I absolutly agree. I also have thought about that feature once again and you convinced me. Please count my vote as +1 now.

The only issue I still have is the right location for those buttons. I think there could be a better dialog window for those buttons. Maybe we add one dialog window for operator options where we could then place those like we have for task instances.

**Comment by feng-tao** [comment]
thanks @feluelle . @Fokko , let us know if you still have concern, @milton0825 ,  I will take a closer look at the actual implementation tonight.

File: airflow/models/baseoperator.py [hunk 1]
```diff
@@ -973,6 +986,25 @@ class BaseOperatorLink:
...
+        """
+        pass
+
     @abstractmethod
     def get_link(self, operator, dttm):
+        # type: (BaseOperator, datetime) -> str
+        """
+        Link to external system.
+
+        :param operator: airflow operator
```
Comment (feng-tao) [inline airflow/models/baseoperator.py:1006]: it would be good to specify the type for arguments as well.

File: tests/plugins/test_plugin.py [hunk 2]
```diff
@@ -100,6 +114,10 @@ class AirflowTestPlugin(AirflowPlugin):
     appbuilder_views = [v_appbuilder_package]
     appbuilder_menu_items = [appbuilder_mitem]
     stat_name_handler = staticmethod(stat_name_dummy_handler)
+    global_operator_extra_links = [
+        GoogleLink(),
```
Comment (feng-tao) [inline tests/plugins/test_plugin.py]: how about just put http://Airflow.apache.org as AirflowLink? Not sure if we want to put google Link or facebook Link in our repo?

File: docs/plugins.rst [hunk 3]
```diff
@@ -202,6 +212,13 @@ definitions in Airflow.
     def stat_name_dummy_handler(stat_name):
         return stat_name
 
+    # A global operator extra link that redirect you to google
+    class GoogleLink(BaseOperatorLink):
```
Comment (feng-tao) [inline docs/plugins.rst]: same. It is not intuitive for user why using this feature with the google link in the doc. Maybe using the kibana log example or qubole url example?

**Review by feng-tao — COMMENTED** [review]
code lgtm, but just minor nits on make it more obvious for users on why or when they should use this feature. Let's also wait for Fokko's input.

File: tests/plugins/test_plugin.py (hunk 2 above)
Comment (feng-tao) [reply tests/plugins/test_plugin.py]: or Kibana log link? Without the internal Lyft knowledge, it isn't very intuitive to user on why we want to put extra link for all the operators.

**Comment by kunalyogenshah** [comment]
Quick question about this implementation. Could you help provide an example usage for let's say Kibana logs that require arguments to be passed to it. For example, if I want to add a global Kibana link for all tasks, I need to be able to specify something like a log ID to the URL to redirect to the right place. Does this implementation allow that?

**Comment by milton0825** [comment]
@kunalyogenshah so currently the operator instance and the `execution_time` are exposed to you like in: https://github.com/apache/airflow/pull/5094/files#diff-2fdb519812514e29589938d977d68ea0R94 and you should have the information (dag_id, task_id,...) to formulate the URL with granularity to each task_instance. If you need information such as log id that is out of the scope of airflow, you can do whatever you need in the `get_link` to get that information.

**Comment by kunalyogenshah** [comment]
@milton0825 Got it. Constructing it this way is certainly doable. But somehow it does seem like introducing the Kibana log URLs in a more central way like the Download log links / View log links could be more semantically appropriate than adding them through global extra links? Specially if you consider adding try number to the mix, in which case a Kibana global should return multi-links instead of one link. Just a thought. + @KevinYang21

**Comment by KevinYang21** [comment]
Nice idea to make it pluggable. We do also have the need to put a Kibana link on the dialog but as @kunalyogenshah mentioned, for task logs I think it makes more sense to provide different links for different `try_number`. Possible that we can make it so flexible that we can have plugin section just like the download log section?

**Comment by milton0825** [comment]
@KevinYang21 @kunalyogenshah I think it totally make sense to include the try-number into the scope.   

I am thinking about addressing that in a separate PR to limit the scope of changes in this PR and here is the jira to track that: https://issues.apache.org/jira/browse/AIRFLOW-4398

File: airflow/models/baseoperator.py (hunk 1 above)
Comment (milton0825) [reply airflow/models/baseoperator.py:1006]: I have a type hint in: https://github.com/apache/airflow/pull/5094/files/8be65327e8cd1f4cf7dedcb90b8eacc4fa5f635d#diff-b5ec97b8301ab688b7e045426f24d485R1002

File: tests/plugins/test_plugin.py (hunk 2 above)
Comment (milton0825) [reply tests/plugins/test_plugin.py]: This is for testing purpose so I tried to make it simple.

**Comment by feng-tao** [comment]
@milton0825 , is there a reason we don't make all the fields available(https://github.com/apache/airflow/blob/master/airflow/models/taskinstance.py#L126-L144) or at least most of the fields? I would imagine that adding pool / priority_weight to the task will be useful as well.

**Comment by milton0825** [comment]
@feng-tao we exposed the operator so all the following fields should be available to user:
https://github.com/apache/airflow/blob/8be65327e8cd1f4cf7dedcb90b8eacc4fa5f635d/airflow/models/baseoperator.py#L241-L274

**Comment by KevinYang21** [comment]
@feng-tao Took a quick look at the code and I think that's because we start to construct the links from operator instead of from task instance. But since we already have the `execution_date` we can always construct the TI just like other endpoints and use whatever info is inside.

While getting the TI fields is easy, my thinking is still more on the form of the link. The fact that we can just construct a link button right now is probly not going to look very nice when we need some other format, like when we want the download log look for our kibana log links.

EDIT: (hit comment button too soon) I'm not trying to scope out this PR, I think there might be use cases that a link button is enough. If it is not trivial to create such a framework for more complex pluggable components in the dialog it makes sense to start another PR around that.

**Comment by feng-tao** [comment]
yeah, FORM/Design/UI is always a tough question....

BTW @KevinYang21 , is yrqls21@gmail.com your email ?

**Comment by KevinYang21** [comment]
@feng-tao yes it is.

**Comment by feng-tao** [comment]
@KevinYang21 , not sure how to DM you, but you may want to check your email :)

File: docs/plugins.rst (hunk 3 above)
Comment (milton0825) [reply docs/plugins.rst]: Updated the docs to give user a link to S3 logs example

**Comment by milton0825** [comment]
@feng-tao this is ready for another look
tests passed on:  https://travis-ci.org/milton0825/airflow/builds/523800135

**Review by feng-tao — APPROVED** [review]
lgtm

**Comment by milton0825** [comment]
@Fokko let us know if you have any other concerns.

**Pull request merged**